distributed/safetensor_utils.py

```diff
@@ -0,0 +1,325 @@
+        for k in f.keys():
+            tensors[k] = f.get_tensor(k)
+    logger.info(f"Loaded {len(tensors)} tensors from {full_path}")
+    return tensors
+
+
+def update_state_dict(
+    state_dict: Dict[str, torch.Tensor],
+    checkpoint: Dict[str, torch.Tensor],
+    weight_map: Dict[str, str],
+    new_to_old_keymap: Dict[str, str],
+    file: str,
+    updated_states: Set[str],
+    device: torch.device,
+):  
+    count_dtensors_loaded = 0
+    for param, file_with_param in weight_map.items():
+        if file_with_param == file and param in state_dict:
+            model_param = (
+                "output.weight" if param == "output.weight" else f"model.{param}"
+            )
+            old_param = new_to_old_keymap.get(model_param)
+
+            if old_param not in checkpoint:
+                logger.warning(f"Missing {old_param} in checkpoint")
+                continue
+
+            checkpoint_tensor = checkpoint[old_param]
+            stage_tensor = state_dict[param]
+            
+            # stage_is_dtensor = is_dtensor(stage_tensor)
+            # logger.info(f"cme DType Check: {param=}, {stage_is_dtensor=}, {checkpoint_tensor.dtype=}, {stage_tensor.dtype=}")
+            
+            checkpoint_tensor = compare_and_reverse(stage_tensor, checkpoint_tensor)
+            
+            # here we need to check if the tensor is a DTensor and if so, adjust the 
+            # shape and placement to match the model DTensor.  
+            if is_dtensor(stage_tensor):
+                model_tensor = load_into_dtensor(checkpoint_tensor, stage_tensor)
+                # logger.info(f"DTensor: Loaded {param} into {model_tensor=}")
+                
+                state_dict[param] = model_tensor
+                count_dtensors_loaded += 1
+                
+            else:
+                # regular tensor, just update directly
+                checkpoint_tensor = checkpoint_tensor.to(device)
+                state_dict[param] = checkpoint_tensor
+
+            state_dict[param] = state_dict[param].to(checkpoint_tensor.dtype)
+            
+            assert state_dict[param].dtype == checkpoint_tensor.dtype
+            assert state_dict[param].dtype == torch.float16, f"{param} dtype is not fp16"
+            
+            # log_tensor_info(param, state_dict[param])
+            # logger.info(f"Loaded {param} from {file}")
+            updated_states.add(param)
+    # logger.info(f"Count of loaded DTensors: {count_dtensors_loaded}")
+    
+    
+def format_tensor_info(tensor: torch.Tensor) -> str:
+    return f"Shape: {tensor.shape}, Dtype: {tensor.dtype}, Device: {tensor.device}"
+
+
+def clean_cache_keys(input_set: Set[str]) -> Set[str]:
+    """Remove cache, freqs, mask params from checkpoint update set...we expect these to be generated """
+    return {
+        item
+        for item in input_set
+        if not (item.endswith("cache") or item in ["freqs_cis", "causal_mask"])
+    }
+
+def handle_missing_keys(
```

Comment:
nit: comment for this function?

Reply:
ah good catch - I try to doc-string every function.  Have added.